lib/logster/redis_store.rb

```diff
@@ -2,6 +2,88 @@
 require 'logster/base_store'
 
 module Logster
+  class RedisRateLimiter
+    BUCKETS = 6
+
+    attr_reader :key, :callback_key
+
+    def initialize(redis, severities, limit, duration, callback = nil)
+      @redis = use_raw_connection? ? Logster.config.redis_raw_connection : redis
+      @severities = severities
+      @limit = limit
+      @duration = duration
+      @callback = callback
+
+      # "_LOGSTER_RATE_LIMIT:012:20:30"
+      # Triggers callback when log levels of :debug, :info and :warn occurs 20 times within 30 secs
+      @key = "#{key_prefix}#{@severities.join("")}:#{@limit}:#{@duration}"
+      @callback_key = "#{@key}:callback_triggered"
+      @bucket_range = @duration / BUCKETS
+    end
+
+    def check(severity)
+      return unless @severities.include?(severity)
+      time = Time.now.to_i
+      num = bucket_number(time)
+      redis_key = "#{@key}:#{num}"
+
+      current_rate = @redis.eval <<-LUA
+        local bucket_number = #{num}
+        local bucket_count = redis.call("INCR", "#{redis_key}")
+
+        if bucket_count == 1 then
+          redis.call("EXPIRE", "#{redis_key}", "#{bucket_expiry(time)}")
+          redis.call("DEL", "#{callback_key}")
+        end
+
+        local function retrieve_rate ()
+          local sum = 0
+          local values = redis.call("MGET", #{mget_keys(num)})
+          for index, value in ipairs(values) do
+            if value ~= false then sum = sum + value end
+          end
+          return sum
+        end
+
+        return (retrieve_rate() + bucket_count)
+      LUA
```

Comment:
@SamSaffron Updated the LUA script :) I removed the need for a set which made it alot simpler now.